pkg/controller/ibmblockcsi/ibmblockcsi_controller.go

```diff
@@ -228,8 +255,69 @@ func (r *ReconcileIBMBlockCSI) Reconcile(request reconcile.Request) (reconcile.R
 	return reconcile.Result{}, nil
 }
 
+func (r *ReconcileIBMBlockCSI) addFinalizerIfNotPresent(instance *ibmblockcsi.IBMBlockCSI) error {
+	logger := log.WithName("addFinalizer")
+
+	accessor, finalizerName, err := r.getAccessorAndFinalizerName(instance)
+	if err != nil {
+		return err
+	}
+
+	if !util.Contains(accessor.GetFinalizers(), finalizerName) {
+		logger.Info("adding", "finalizer", finalizerName, "on", accessor.GetName())
+		accessor.SetFinalizers(append(accessor.GetFinalizers(), finalizerName))
+
+		if err := r.client.Update(context.TODO(), instance.Unwrap()); err != nil {
+			logger.Error(err, "failed to add", "finalizer", finalizerName, "on", accessor.GetName())
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *ReconcileIBMBlockCSI) hasFinalizer(instance *ibmblockcsi.IBMBlockCSI) (bool, error) {
+	accessor, finalizerName, err := r.getAccessorAndFinalizerName(instance)
+	if err != nil {
+		return false, err
+	}
+
+	if util.Contains(accessor.GetFinalizers(), finalizerName) {
+		return true, nil
+	}
+	return false, nil
+}
+
+func (r *ReconcileIBMBlockCSI) removeFinalizer(instance *ibmblockcsi.IBMBlockCSI) error {
+	logger := log.WithName("removeFinalizer")
+
+	accessor, finalizerName, err := r.getAccessorAndFinalizerName(instance)
+	if err != nil {
+		return err
+	}
+
+	accessor.SetFinalizers(util.Remove(accessor.GetFinalizers(), finalizerName))
+	if err := r.client.Update(context.TODO(), instance.Unwrap()); err != nil {
+		logger.Error(err, "failed to remove", "finalizer", finalizerName, "from", accessor.GetName())
+		return err
+	}
+	return nil
+}
+
+func (r *ReconcileIBMBlockCSI) getAccessorAndFinalizerName(instance *ibmblockcsi.IBMBlockCSI) (metav1.Object, string, error) {
+	logger := log.WithName("getAccessorAndFinalizerName")
+	lowercaseKind := strings.ToLower(instance.GetObjectKind().GroupVersionKind().Kind)
+	finalizerName := fmt.Sprintf("%s.%s", lowercaseKind, oconfig.APIGroup)
+
+	accessor, err := meta.Accessor(instance)
+	if err != nil {
+		logger.Error(err, "failed to get meta information of instance")
+		return nil, "", err
+	}
+	return accessor, finalizerName, nil
```

Comment:
Is it considered good practice to return a tuple of three elements in Go? IMHO, when we want to return more than two elements, we usually wrap them with a dictionary/object since we don't want to "remember" the order of the elements whenever we call this function.

Reply:
what is special about the number "two" in this rule? and how would you name the object here?